# Greetings

###### Hello my name is Swaraj Bari, from india and i am sending this project documentation. This is actually my first hands on project so I currently don't know how I should go about this document, so I am just documenting in a way that felt nice.

# Step- 1 Data Cleaning and Organizing

###### In my endeavor to solve this challenge, this was frankly the hardest and most time taking step

FIRSTLY, I SAW ALL THE EXCEL FILES AND GOT MYSELF FAMILIARIZED BY IT, I BRAINSTORMED ON WHAT I SHOULD DO ABOUT IT AND DECIDED ON A TYPE OF LAYOUT I WANTED FOR THIS DATA.

AFTER THAT THE NEXT STEP WAS IMPLEMENTING IT, I SEARCHED FOR A LOT OF FUNCTIONS IN THE PANDAS OFFICIAL DOCUMENTATION, BUT COULD NOT EFFECTIVELY THOSE FUNCTIONS

SO I DECIDED TO CREATE MY OWN CUSTOM FUNCTIONS TO DO THE TASK AS FOLLOWS:

In [1]:
import pandas as pd
import numpy as np
import datetime


def passage_type(df):
    return df[0][0]

def row_index(df):
    index_req = 0
    for element in df[0][:5]:
        if type(element)==str:
            if 'quest' in element.lower():
                return index_req
        index_req += 1
    return

def date_time_search(df, row_inde):
    '''
    Return the result in (row, col) format, colllect it as such tuple form in only one variable
    also put row in df.iloc[]
    and col in df[]'''
    for i in range(row_inde):
        col=0
        row = i
        for o in df.iloc[i]:
            if type(o)==datetime.datetime:
                return (row, col)
            
            col += 1
    return

def no_of_col(df):
    return len(df.columns)

def no_of_rows(df):
    return len(df)

def student_date_dictionary(df, rowcol, row_inde):
    '''
    rowcol is a tuple from date_time_searc() function
    row_inde is output of row_index() function'''
    date = []
    for i in df.iloc[rowcol[0]][rowcol[1]:]:
        date.append(i)
        
    students = []    
    for j in df.iloc[row_inde][rowcol[1]:]:
        students.append(j)
    
    diction = dict(zip(students, date))
    del date
    del students
    
    return diction

def clean_datframe(df):

    row_inde = row_index(df)
    
    rowcol = date_time_search(df, row_inde)
    total_col = no_of_col(df)
    dates = student_date_dictionary(df, rowcol, row_inde)
    df2 = df.iloc[row_inde+1:].copy()
    df2.columns = df.iloc[row_inde]

    return df2

def data_creation(df, rowcol):
    students = df.columns[rowcol[1]:]
    
    data = {'student_id':[i for i in students for j in range(len(df))],
           'student_answer':[j for i in range(rowcol[1], len(df.columns)) for j in df.iloc[:,i]],
           'correct_answer':[j for i in students for j in df['Answer']],
           'question_no':[j for i in students for j in df['Question']],
           'question_type':[j for i in students for j in df['Question Type']]}
    return data

def dataframing(data, sheet_name, passage_ttpe, test_type):
    df = pd.DataFrame(data)
    df['age'] = df['student_id'].apply(age_calculator)
    df['test_type'] = test_type
    df['passage_name'] = sheet_name
    df['passage_type'] = passage_ttpe 
    return df

def student_birthday():
    dd = pd.read_excel(r"D:\Machine_Learning\Dicey Tech\DiceyTech x Target Tuition (Official)\student profiles.xlsx")
    data_dict = dd.set_index('Student ID')['Birthday'].to_dict()
    #print(data_dict)
    del dd
    return data_dict

def age_calculator(x):
    '''
    dates student_dates_dictionary() function se
    data_dict () us function se
    df humara new generated df hai
    '''
    global data_dict
    global dates
    try:
        age = dates[x]-data_dict[x]
        age = age.days/365
        return age
    except:
        return np.NaN
    
    
    
    
    
def data_processor_proto2(df,sheets,test_type):
    
    sheets = sheets
    dataframes =[]
    for sheet in sheets:
        df1 = df[sheet].copy()
        row_inde = row_index(df1)
        passage = passage_type(df1)
        rowcol = date_time_search(df1, row_inde)
        global dates
        dates = student_date_dictionary(df1,rowcol, row_inde)
        df1 = clean_datframe(df1)
        data = data_creation(df1, rowcol)
        del df1
        df2 = dataframing(data, sheet, passage, test_type)
        dataframes.append(df2)
    
    return dataframes

In [2]:
# USING PANDAS READ THE worksheet.xlsx KEEPING header=None and sheet_name = None
df1 = pd.read_excel(r"D:\Machine_Learning\Dicey Tech\DiceyTech x Target Tuition (Official)\Worksheets\worksheet data.xlsx", sheet_name=None, header=None)

In [3]:
#  This returned me a dictionary of dataframes with the sheet name as key and thier corresponding dataframes as value

In [4]:
# After that a fetched all the names of the sheets, using the following three lines of code
sheets = list(df1.keys())
#  had to remove these two because they were empty and also there was no text data on both of them
sheets.remove('robin -happened')
sheets.remove('robin -considered')
sheets

['anne',
 '1984',
 'alice',
 'baskervilles',
 'black beauty',
 'brazilian',
 'charlottes',
 'court life',
 'DORIAN',
 'dreams of a giant life',
 'economics',
 'fauntleroy',
 'frankenstein',
 'friendship',
 'frost',
 'gatsby',
 'gibson',
 'growth',
 'gulliver',
 'homing pigeon',
 'honeybees',
 'huck',
 'i have a dream',
 'jabberwocky',
 'JANE EYRE',
 'jekyll',
 'lion',
 'little women',
 'lochinvar',
 'lost treasure',
 'magic city',
 'marathon',
 'matilda',
 'meg',
 'no enemies',
 'nothing gold',
 'NOTRE DAME',
 'oliver twist -master',
 'oliver twist -room',
 'sense',
 'patagonia',
 'peter',
 'rebecca',
 'sea fever',
 'secret garden',
 'shiba',
 'snowy river',
 'sonnet 18',
 'stopping by',
 'this is going to hurt',
 'THOSE WINTER',
 'tom sawyer',
 'treasure island',
 'tyger',
 'umami',
 'van',
 'voyage',
 'war of the worlds',
 'when u r old',
 'wind in the willows',
 'wizard']

In [5]:
#  after that I ran the function i created that was data_processor_proto2
df1 = data_processor_proto2(df1,sheets, test_type='worksheet')
df1

[     student_id student_answer correct_answer question_no question_type  age  \
 0             8              C              C           1             M  NaN   
 1             8              C              C           2             M  NaN   
 2             8              C              D           3             H  NaN   
 3             8              D              C           4        M, T/F  NaN   
 4             8              D              D           5             H  NaN   
 ..          ...            ...            ...         ...           ...  ...   
 283          45              D              D          12             G  NaN   
 284          45              C              C          13             M  NaN   
 285          45              A              A          14             M  NaN   
 286          45              D              D          15             M  NaN   
 287          45              C              C        (16)             P  NaN   
 
      test_type passage_na

THESE ARE THE LIST OF ALL THE DATFRAMES I WANTED IN THE SPECIFIC LAYOUT I WANTED

THE FEATURES I ADDED AT THIS POINT VIA MY CUSTOM FUNCTIONS ARE age, test_type(i.e., worksheet or mock), passage_name, and passage_type

THEN I DID THE SAME FOR THE Mock Exam Data (Labelled).xlsx


In [6]:
df2 = pd.read_excel(r"D:\Machine_Learning\Dicey Tech\DiceyTech x Target Tuition (Official)\Mock Exams\Mock Exam Data (Labelled).xlsx", sheet_name=None, header=None)

In [7]:
mock_sheets = list(df2.keys())
mock_sheets

['bluebird',
 'dorian gray',
 'jane eyre',
 'notre-dame',
 'metamorphosis',
 'sleepy hollow',
 'the hobbit',
 'those winter']

In [8]:
df2 = data_processor_proto2(df2, mock_sheets, test_type='mock')
df2

[     student_id student_answer correct_answer  question_no question_type  age  \
 0            61              C              D            1             H  NaN   
 1            61              D              D            2             M  NaN   
 2            61              B              B            3             G  NaN   
 3            61              C              D            4             M  NaN   
 4            61              B              B            5             M  NaN   
 ..          ...            ...            ...          ...           ...  ...   
 795           1              A              A           21             H  NaN   
 796           1              A              D           22             H  NaN   
 797           1              B              D           23             H  NaN   
 798           1              A              A           24             G  NaN   
 799           1              C              C           25             H  NaN   
 
     test_type

AFTER THIS I ADDED BOTH THE LISTS OF DATAFRAMES

In [9]:
df1 = df1 + df2
df1

[     student_id student_answer correct_answer question_no question_type  age  \
 0             8              C              C           1             M  NaN   
 1             8              C              C           2             M  NaN   
 2             8              C              D           3             H  NaN   
 3             8              D              C           4        M, T/F  NaN   
 4             8              D              D           5             H  NaN   
 ..          ...            ...            ...         ...           ...  ...   
 283          45              D              D          12             G  NaN   
 284          45              C              C          13             M  NaN   
 285          45              A              A          14             M  NaN   
 286          45              D              D          15             M  NaN   
 287          45              C              C        (16)             P  NaN   
 
      test_type passage_na

THEN I CONCATINATED THE DATAFRAMES IN THE LIST TO CREATE A MASTER DATAFRAME AND EXPORTED IT TO A EXCEL FILE

In [10]:
combined_df = pd.concat(df1, ignore_index=True)
combined_df

student_id student_answer correct_answer question_no question_type  \
0             8.0              C              C           1             M   
1             8.0              C              C           2             M   
2             8.0              C              D           3             H   
3             8.0              D              C           4        M, T/F   
4             8.0              D              D           5             H   
...           ...            ...            ...         ...           ...   
11829        33.0              A              A          13             H   
11830        33.0              C              C          14             V   
11831        33.0              D              D          15             H   
11832        33.0              A              A          16             H   
11833        33.0              A              A          17             M   

       age  test_type  passage_name passage_type  
0      NaN  worksheet          anne      fiction  
1      NaN  worksheet          anne      fiction  
2      NaN  worksheet          anne      fiction  
3      NaN  worksheet          anne      fiction  
4      NaN  worksheet          anne      fiction  
...    ...        ...           ...          ...  
11829  NaN       mock  those winter         poem  
11830  NaN       mock  those winter         poem  
11831  NaN       mock  those winter         poem  
11832  NaN       mock  those winter         poem  
11833  NaN       mock  those winter         poem  

[11834 rows x 9 columns]

THIS IS THE WHOLE DATA WHICH NOW HAS 11834 ROWS 

In [ ]:
combined_df.to_csv('organized_data.csv', index=False)

# Step 2- EXTRACTING FEATURES FROM THE PASSAGE TEXT AND QUESTION ANSWERS

###### MY FIRST THOUGHT WAS TO JUST ADD NEW COLUMNS TO THE ABOVE DATFRAME WHICH WOULD BE THE 
1) PASSAGE TEXT
2) QUESTION
3) OPTION A
4) OPTION B
5) OPTION C
6) OPTION D

###### THIS WOULD BE ACCORDINGLY AND THEN SEPERATE THESE COLUMNS IN A NEW DATAFRAME AND DO THE STANDARD NLP PROCESSES LIKE LOWERCASING FOLLOWED WITH REMOVING PUNCTUATIONS, LEMMATIZATION, TOKENIZATION AND THE APPLY A TECHNIQUE LIKE THE BAG OF WORDS, OF TF-IDF VECTOR.

###### BUT THEN I THOUGHT THAT THESE TECHNIQUES WILL CREATE A LOT OF FEATURES AND MY SYSTEM WILL NOT BE ABLE TO HANDLE THE COMPUTATION SO I CHABGE MY APPROACH AND DECIDED FOR A DIFFERENT SET OF FEATURES

###### i USED THE SPACY LIBRARY TO DO THE STANDARD NLP PROCESSES LIKE LOWERCASING FOLLOWED WITH REMOVING PUNCTUATIONS, LEMMATIZATION AND THE I FOUND THE POS TAGS (FINE POS TAGGING) AND THEIR DEPENDENCY PARSINGS FOR EACH TEXT AND USED THE COUNTER METHOD FROM COLLECTIONS LIBRARY TO COUNT THE NO OF OCCURENCED OF EACH TAGS AND DEPENDENCIES FOR EVERY DIFFERENT TEXT, QUESTION AND ANSWWERS RESPECTIVELY , WHICH I WOULD USE AS FEATURES.


###### ON TOP OF THESE FEATURES I ALSO CUSTOM CREATED SOME FEATURES FOR EVERY DIFFERENT TEXT, QUESTION AND ANSWWERS RESPECTIVELY , VIZ,  
1) LENGTH OF PASSAGE 
2) LENGTH OF QUESTION
3) NO OF WORDS IN THE PASSAGE
4) NO OF WORDS IN THE QUESTIONS
5) NO OF WORDS IN OPTION A
6) SIMILIARLY NO OF WORD IN OPTION B, C AND D
7) NO OF SENTENCES IN THE PASSAGE, QUSTION, OPTION A,B,C,D 
8) AVG LENGTH OF SENTENCES IN THE PASSAGE, QUSTION, OPTION A,B,C,D 
9) AVG NO OF WORDS IN SENTENCES IN THE PASSAGE, QUSTION, OPTION A,B,C,D 

In [11]:
import spacy
import os
import sys
from collections import Counter
nlp = spacy.load('en_core_web_sm')

In [13]:
#  WENT TO THE FOLDER WHERE I ALREADY HAD THE THE TEXT DAT, 
# SEPEREATED BETWEEN PASSAGE TEXT AND THE QUESTION WERE SEPERATED IN A JSON FILE I CREATED
folder = r'D:\Machine_Learning\Dicey Tech\Project\data\Workshet'
file_list = os.listdir(folder)
print(file_list)

['1984.txt', 'alice in woonderland.txt', 'anne of green gables.txt', 'black beauty.txt', 'brazillian.txt', 'charlrttes web.txt', 'court life from within.txt', 'dorian.txt', 'dream of gaint life.txt', 'economics for children.txt', 'fauntleroy.txt', 'frankestien.txt', 'friendship.txt', 'frost.txt', 'gatsby.txt', 'gibson.txt', 'growth mindset.txt', 'gulivers travels.txt', 'homing piegon.txt', 'honey bees.txt', 'hound of baskerville.txt', 'huckleberry finn.txt', 'i have a dream.txt', 'jaberwocky.txt', 'jane eyre.txt', 'jekyll.txt', 'lion.txt', 'little women.txt', 'lochinvar.txt', 'lost in the woods.txt', 'lost treasure.txt', 'magic city.txt', 'marathon.txt', 'matilda.txt', 'meg of mysterious mountain.txt', 'New Text Document.txt', 'no enemies.txt', 'northern lights.txt', 'note darme of paris.txt', 'nothing gold can stay.txt', 'oliver twist master.txt', 'oliver twist room.txt', 'patagonia.txt', 'peterpan.txt', 'pollyanna.txt', 'rebecca of sunnybrook farm.txt', 'sea fever.txt', 'secret garde

In [25]:
#  custom function i created to process the passage text data
import re
import string
punct = string.punctuation


def nlp_processor_for_passage_text(file_list, punct):

    dictionary_passage = dict()
    for file_name in file_list:
        file = open(f'D:\\Machine_Learning\\Dicey Tech\\Project\\data\\Workshet\\{file_name}', encoding='utf-8')
        data = file.read()
        file.close()
        data=data.replace('\n',' ').lower()
        data = re.sub(f'[{punct}]', '', data)
        passge_length = len(data)
        doc = nlp(data)

        no_of_words = len(doc)
        no_of_entities = len(doc.ents)

        no_of_sentence = 0
        no_of_words = 0
        len_of_sent = 0
        for sents in doc.sents:
            no_of_sentence += 1
            no_of_words += len(sents)
            len_of_sent += len(sents.text)
        #     print(len(sents))
        if no_of_sentence == 0:
            avg_length = 0
            avg_words = 0
        else:
            avg_words = no_of_words/no_of_sentence
            avg_length = len_of_sent/no_of_sentence




        lst = []
        for word in doc:
            tag = word.tag_
            lst.append(tag)
            lst.append(word.dep_)
        #     print(word,'----->', tag,'------->', spacy.explain(tag), '------->', word.dep_)

        features = Counter(lst)
        file_title = file_name.split('.')[0]
        dictionary_passage[f'{file_title}'] = features
        dictionary_passage[f'{file_title}']['passage_length'] = passge_length
        dictionary_passage[f'{file_title}']['no_of_words'] = no_of_words
        dictionary_passage[f'{file_title}']['no_of_sentances'] = no_of_sentence
        dictionary_passage[f'{file_title}']['avg_len_of_sentenc'] = avg_length
        dictionary_passage[f'{file_title}']['avg_no_of_words_per_sentence'] = avg_words
        dictionary_passage[f'{file_title}']['no_of_entities'] = no_of_entities

        del features

    return dictionary_passage

In [26]:
dictionary_passage = nlp_processor_for_passage_text(file_list, punct)
dictionary_passage
# these are all the features extracted from th passage text

{'1984': Counter({'CD': 6,
          'ROOT': 3,
          'IN': 89,
          'prep': 74,
          'NNP': 15,
          'compound': 18,
          'pobj': 72,
          'PRP': 27,
          'nsubj': 47,
          'VBD': 55,
          'ccomp': 20,
          'DT': 86,
          'det': 87,
          'JJ': 40,
          'amod': 36,
          'NN': 121,
          'attr': 13,
          'CC': 18,
          'cc': 19,
          'NNS': 21,
          'aux': 20,
          'VBG': 15,
          'conj': 30,
          'nummod': 3,
          'dobj': 23,
          'PRP$': 8,
          'poss': 9,
          'TO': 5,
          'VB': 10,
          'acl': 6,
          'RB': 38,
          'advmod': 43,
          'neg': 1,
          'xcomp': 3,
          'pcomp': 4,
          'RP': 10,
          'prt': 10,
          'VBN': 22,
          'nsubjpass': 9,
          'auxpass': 10,
          'relcl': 15,
          'JJR': 1,
          'quantmod': 1,
          'npadvmod': 4,
          'JJS': 1,
          'WP': 1,
   

In [24]:
#  imported this to json to save this
import json
with open('passage_feature.json', 'w') as f:
    json.dump(dictionary_passage, f)

None


In [28]:
#  Now for the question and answers
import json
with open('question_answer.json') as f:
    py_di = json.load(f)
    
py_di
# i had already created a dictionary of question and answers and stored in in a json file the contents of the dictionary are as follows

{'mock': {'bluebird': {'1': ['The first two lines are repeated four times throughout the poem. What does this suggest about the bluebird?option 1 he is determined. option 2 he is a constant bother to the poet. option 3 he is tough. option 4 he is monotonous',
    'Option 1 only',
    'Option 2 only',
    'Options 1 and 4 only',
    'Options 1, 2 and 3 only'],
   '2': ['In the stanza 1, the poet says that they are ‘too tough’ for the bluebird. What does this suggest the poet feels?',
    'Letting the bluebird out would hurt the poet',
    'The bluebird is closely matched to the poet in strength',
    'The bluebird is stronger than the poet',
    'Letting the bluebird out would mean the poet is too weak'],
   '3': ['The poet says ‘stay in there’ to the bluebird. This is an example of an imperative clause. What is an imperative clause?',
    'An important statement',
    'An order or instruction',
    'A request',
    'An exclamation'],
   '4': ['How does the poet stop the bluebird from g

NOW TO PROCESS THIS DATA I CREATED A NEW CUSTOM FUNCTION FOR TIS AND RAN IT AND THEN STORED THE RESULLT IN TO A JSON FILE

In [29]:
def nlp_processor_question_answer(data):
    
    data=data.replace('\n',' ').lower()
    data = re.sub(f'[{punct}]', '', data)
    passge_length = len(data)
    doc = nlp(data)
    no_of_words = len(doc)
    no_of_entities = len(doc.ents)
    lst = []
    for word in doc:
        tag = word.tag_
        lst.append(tag)
        lst.append(word.dep_)
    #     print(word,'----->', tag,'------->', spacy.explain(tag), '------->', word.dep_)

    features = Counter(lst)
    features['question_length'] = passge_length
    features['no_of_words'] = no_of_words
    features['no_of_entities'] = no_of_entities
    return features

In [30]:

for test_type in py_di:
    
    for passage in py_di[test_type]:
        for question_no in py_di[test_type][passage]:
            for i in range(len(py_di[test_type][passage][question_no])):
                features = nlp_processor_question_answer(py_di[test_type][passage][question_no][i])
                py_di[test_type][passage][question_no][i] = features
                


{}

In [31]:
py_di

{'mock': {'bluebird': {'1': [Counter({'DT': 6,
             'det': 5,
             'JJ': 5,
             'amod': 4,
             'CD': 6,
             'nummod': 3,
             'NNS': 2,
             'nsubjpass': 1,
             'VBP': 1,
             'auxpass': 1,
             'VBN': 1,
             'ccomp': 4,
             'npadvmod': 3,
             'IN': 3,
             'prep': 3,
             'NN': 7,
             'pobj': 3,
             'WP': 1,
             'dobj': 1,
             'VBZ': 5,
             'aux': 1,
             'nsubj': 5,
             'VB': 1,
             'PRP': 4,
             'attr': 3,
             'ROOT': 2,
             'compound': 1,
             'punct': 1,
             'acomp': 1,
             'question_length': 222,
             'no_of_words': 42,
             'no_of_entities': 7}),
    Counter({'NN': 1,
             'ROOT': 1,
             'CD': 1,
             'nummod': 1,
             'RB': 1,
             'advmod': 1,
             'question_length':

In [32]:
#  storing this to a json file
with open('question_answer_feature.json', 'w') as f:
    json.dump(py_di,f)

# Step-3 

I THEN DID A LOT OF MANIPULATION ON THE DATA I HAD TO CREATE A MASTER DATFRAME AND EXPORTED IT AS AN EXCEL FILE,
THIS IS THE FINAL DATA I NOW HAVE FOR MODELLING.

THE CODE OF THIS STEP ARE A BIT OF A MESS HONESTLY SO I AM ONLY SHOWING THE FINAL DATFRAME

In [33]:
DATAFRAME = pd.read_excel('complete_data.xlsx')
DATAFRAME

student_id student_answer correct_answer  question_no question_type  \
0              8              C              C            1             M   
1             15              C              C            1             M   
2             22              C              C            1             M   
3              2              C              C            1             M   
4              6              C              C            1             M   
...          ...            ...            ...          ...           ...   
6721          77              A              C           13             M   
6722          74              B              C           13             M   
6723          76              C              C           13             M   
6724           1              C              C           13             M   
6725          51              A              C           13             M   

            age  test_type          passage_name passage_type  passage_NNP  \
0      9.402740  worksheet  anne of green gables      fiction         33.0   
1      9.112329  worksheet  anne of green gables      fiction         33.0   
2      9.260274  worksheet  anne of green gables      fiction         33.0   
3     10.084932  worksheet  anne of green gables      fiction         33.0   
4     10.046575  worksheet  anne of green gables      fiction         33.0   
...         ...        ...                   ...          ...          ...   
6721  10.454795       mock         sleepy hollow      fiction         14.0   
6722  10.564384       mock         sleepy hollow      fiction         14.0   
6723  10.550685       mock         sleepy hollow      fiction         14.0   
6724  10.747945       mock         sleepy hollow      fiction         14.0   
6725  10.241096       mock         sleepy hollow      fiction         14.0   

      ...  DFeature_predet  DFeature_parataxis  DFeature_oprd  DFeature_UH  \
0     ...              NaN                 NaN            NaN          NaN   
1     ...              NaN                 NaN            NaN          NaN   
2     ...              NaN                 NaN            NaN          NaN   
3     ...              NaN                 NaN            NaN          NaN   
4     ...              NaN                 NaN            NaN          NaN   
...   ...              ...                 ...            ...          ...   
6721  ...              NaN                 NaN            NaN          NaN   
6722  ...              NaN                 NaN            NaN          NaN   
6723  ...              NaN                 NaN            NaN          NaN   
6724  ...              NaN                 NaN            NaN          NaN   
6725  ...              NaN                 NaN            NaN          NaN   

      DFeature_FW  DFeature_RBS  DFeature_preconj  DFeature_$  DFeature_nmod  \
0             NaN           NaN               NaN         NaN            NaN   
1             NaN           NaN               NaN         NaN            NaN   
2             NaN           NaN               NaN         NaN            NaN   
3             NaN           NaN               NaN         NaN            NaN   
4             NaN           NaN               NaN         NaN            NaN   
...           ...           ...               ...         ...            ...   
6721          NaN           NaN               NaN         NaN            NaN   
6722          NaN           NaN               NaN         NaN            NaN   
6723          NaN           NaN               NaN         NaN            NaN   
6724          NaN           NaN               NaN         NaN            NaN   
6725          NaN           NaN               NaN         NaN            NaN   

      DFeature_appos  
0                NaN  
1                NaN  
2                NaN  
3                NaN  
4                NaN  
...              ...  
6721             NaN  
6722             NaN  
6723             NaN  
6724            

In [37]:
DATAFRAME.columns

Index(['student_id', 'student_answer', 'correct_answer', 'question_no',
       'question_type', 'age', 'test_type', 'passage_name', 'passage_type',
       'passage_NNP',
       ...
       'DFeature_predet', 'DFeature_parataxis', 'DFeature_oprd', 'DFeature_UH',
       'DFeature_FW', 'DFeature_RBS', 'DFeature_preconj', 'DFeature_$',
       'DFeature_nmod', 'DFeature_appos'],
      dtype='object', length=524)

In [38]:
#  we now have only 6726 rows because thre was a lot of data that was missing

I removed the rows in which the correct answer or student answer was option E, since it was a very few occurences,

Then i created a new column called y_prediction which values to 1 if the student_answer mathes the corect answer and otherwise zero and imported the changes to a new file

# Step - 6   Modelling

In [39]:
import pandas as pd
from sklearn.model_selection import train_test_split as tts

from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier

from sklearn.preprocessing import OrdinalEncoder
from sklearn.preprocessing import OneHotEncoder
from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer

from sklearn.metrics import accuracy_score, confusion_matrix

In [40]:
df = pd.read_excel('model_ready_data.xlsx')
df

student_answer  y_prediction  student_id correct_answer  question_no  \
0                 C             1           8              C            1   
1                 C             1          15              C            1   
2                 C             1          22              C            1   
3                 C             1           2              C            1   
4                 C             1           6              C            1   
...             ...           ...         ...            ...          ...   
6311              A             0          77              C           13   
6312              B             0          74              C           13   
6313              C             1          76              C           13   
6314              C             1           1              C           13   
6315              A             0          51              C           13   

     question_type        age  test_type          passage_name passage_type  \
0                M   9.402740  worksheet  anne of green gables      fiction   
1                M   9.112329  worksheet  anne of green gables      fiction   
2                M   9.260274  worksheet  anne of green gables      fiction   
3                M  10.084932  worksheet  anne of green gables      fiction   
4                M  10.046575  worksheet  anne of green gables      fiction   
...            ...        ...        ...                   ...          ...   
6311             M  10.454795       mock         sleepy hollow      fiction   
6312             M  10.564384       mock         sleepy hollow      fiction   
6313             M  10.550685       mock         sleepy hollow      fiction   
6314             M  10.747945       mock         sleepy hollow      fiction   
6315             M  10.241096       mock         sleepy hollow      fiction   

      ...  DFeature_JJS  DFeature_predet  DFeature_parataxis  DFeature_oprd  \
0     ...             0                0                   0              0   
1     ...             0                0                   0              0   
2     ...             0                0                   0              0   
3     ...             0                0                   0              0   
4     ...             0                0                   0              0   
...   ...           ...              ...                 ...            ...   
6311  ...             0                0                   0              0   
6312  ...             0                0                   0              0   
6313  ...             0                0                   0              0   
6314  ...             0                0                   0              0   
6315  ...             0                0                   0              0   

      DFeature_UH  DFeature_FW  DFeature_RBS  DFeature_preconj  DFeature_$  \
0               0            0             0                 0           0   
1               0            0             0                 0           0   
2               0            0             0                 0           0   
3               0            0             0                 0           0   
4               0            0             0                 0           0   
...           ...          ...           ...               ...         ...   
6311            0            0             0                 0           0   
6312            0            0             0                 0           0   
6313            0            0             0                 0           0   
6314            0            0             0                 0           0   
6315            0            0             0                 0           0   

      DFeature_nmod  
0                 0  
1                 0  
2                 0  
3                 0  
4                 0  
...             ...  
6311              0  
6312              0  
6313              0  
6314              0  
631

In [41]:
df.drop(columns='passage_name', inplace=True)
X = df.iloc[:,2:]
y = df.iloc[:,1]
X_train, X_test, y_train, y_test = tts(X,y,test_size=0.2, random_state=2)

In [42]:
X_train

student_id correct_answer  question_no question_type        age  \
5136           2              A            7             M   9.739726   
3466           1              B            9  AssignMeZero   8.649315   
824           15              A           12  AssignMeZero   9.728767   
5932          35              D           23             H  10.882192   
749           24              A           12  AssignMeZero   9.210959   
...          ...            ...          ...           ...        ...   
1099          22              C           10  AssignMeZero   9.298630   
2514          21              B            6  AssignMeZero   9.526027   
3606          20              D            4  AssignMeZero   9.695890   
5704           1              A           14             M   9.745205   
2575          23              D            9  AssignMeZero  10.076712   

      test_type passage_type  passage_NNP  passage_intj  passage_IN  ...  \
5136  worksheet         poem            0             0          12  ...   
3466  worksheet      fiction           18             0          85  ...   
824   worksheet      fiction            5             0          73  ...   
5932       mock         poem            4             1          14  ...   
749   worksheet      fiction           26             5          63  ...   
...         ...          ...          ...           ...         ...  ...   
1099  worksheet         poem            0             0          35  ...   
2514  worksheet      fiction           24             0         162  ...   
3606  worksheet      fiction           30             2          91  ...   
5704       mock         poem            4             1          14  ...   
2575  worksheet      fiction           24             0         162  ...   

      DFeature_JJS  DFeature_predet  DFeature_parataxis  DFeature_oprd  \
5136             0                0                   0              0   
3466             0                0                   0              0   
824              0                0                   0              0   
5932             0                0                   0              0   
749              0                0                   0              0   
...            ...              ...                 ...            ...   
1099             0                0                   0              0   
2514             0                0                   0              0   
3606             0                0                   0              0   
5704             0                0                   0              0   
2575             0                0                   0              0   

      DFeature_UH  DFeature_FW  DFeature_RBS  DFeature_preconj  DFeature_$  \
5136            0            0             0                 0           0   
3466            0            0             0                 0           0   
824             0            0             0                 0           0   
5932            0            0             0                 0           0   
749             0            0             0                 0           0   
...           ...          ...           ...               ...         ...   
1099            0            0             0                 0           0   
2514            0            0             0                 0           0   
3606            0            0             0                 0           0   
5704            0            0             0                 0           0   
2575            0            0             0                 0           0   

      DFeature_nmod  
5136              0  
3466              0  
824               0  
5932              0  
749               0  
...             ...  
1099              0  
2514              0  
3606              0  
5704              0  
2575              0  

[5052 rows x 514 columns]

In [43]:
y_train


5136    1
3466    0
824     0
5932    0
749     0
       ..
1099    0
2514    0
3606    1
5704    0
2575    1
Name: y_prediction, Length: 5052, dtype: int64

##### TRANSFORMER

In [44]:
transformer = ColumnTransformer(transformers=[
    ('tup1', SimpleImputer(add_indicator=True), ['age']),
    ('tup2', OrdinalEncoder(categories=[['AssignMeZero','M', 'H', 'M, T/F','P','G','V' ]]), ['question_type']),
    ('tup3', OneHotEncoder(drop='first', sparse=False), ['correct_answer','test_type','passage_type']),
    ('tup4', OneHotEncoder(sparse=False), ['student_id'])
], remainder='passthrough')

In [45]:
X_train_transformed = transformer.fit_transform(X_train)
X_test_trnasformed = transformer.transform(X_test)

C:\Users\swaraj\AppData\Local\Programs\Python\Python38\lib\site-packages\sklearn\preprocessing\_encoders.py:972: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(
C:\Users\swaraj\AppData\Local\Programs\Python\Python38\lib\site-packages\sklearn\preprocessing\_encoders.py:972: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


##### AFTER TRANFORMATION OF THE DATA, I FIRST RAN 

##### 1) LOGISTIC REGRESSION ON THE DATA, WHICH GAVE ME AN ACCURACY OF 0.5791139240506329 

##### 2) SUPPORT VECTOR MACHINE WITH A SIGMOID KERNEL GAVE AN ACCURACY OF 0.5118670886075949 AND THE OTHRE KERNELS VIZ: RBF, POLY BOTH GAVE A SIMILIAR ACCURACY SCORE

##### 3) I TRIED RANDOM FOREST ON IT TOO WHICH GAVE A SCORE OF 0.6542721518987342, BUT AT THIS POINT I HONESTLY DON'T KNOW ANY MATHS BEHIND BEHIND THIS ALGORITHM, I USED IT BECAUSE IT IS VERY POPULAR, BUT I DON'T UNDERSTAND IT AS IN DEPTH AS I UNDERSTAND THE ABOVE TWO YET

FOR THE PART TWO OF THE CHALLENGE ALL I CHANGED WAS I TOOK THE Y AS correct_answer COLUMN AND APLLIED THE TRANSFROMATIONS ACCORDINGLY, BUT FOR THAT ALSO THE ACCURACY SCORES WERE SIMILAR




I HAVE A FEW MORE IDEAS I WOULD LIKE TO TRY BUT UNFORTUNATELY TODAY IS THE END OF THE TIME FOR THIS CHALLENGE, BUT I HAVE LEARNED A LOT AND HAVE SOME REAL TAKEAWAYS FROM THIS CHALLENGE ALREADY,

I KNOW THAT A MAX ACCURACY SCORE OF AROUND 65 PRECENT IS NOT GOOD ENOUGH TO MAKE ME WIN THE COMPETITION BUT I LEARNED A LOT AND DON'T REGERET MY TIME DEVOTED ON THIS PROJECT

I WILL CONTINUE TO WORK ON THIS AS I KEEP ON LEARNING IN THE FIELD OF ML.

I AM TRULY GRATEFUL TO DICEY TECH AND TARGET TUITION FOR THIS WONDERFUL OPPURTUNITY .
THIS CHALLENGE HAS FUELED MY CURIOSITY OF ML EVEN MORE, SO THANK YOU REALLY.


